addons/mail/static/src/messaging/component/follower_subtype/follower_subtype_tests.js

```diff
@@ -0,0 +1,239 @@
+        document.querySelectorAll('.o_FollowerSubtype_checkbox').length,
+        1,
+        "should have a checkbox"
+    );
+    assert.strictEqual(
+        document.querySelector('.o_FollowerSubtype_label').textContent,
+        "Dummy test",
+        "should have the name of the subtype as label"
+    );
+    assert.ok(
+        document.querySelector('.o_FollowerSubtype_checkbox').checked,
+        "checkbox should be checked as follower subtype is followed"
+    );
+});
+
+QUnit.test('simplest layout of a not followed subtype', async function (assert) {
+    assert.expect(5);
+
+    await this.start();
+    const thread = this.env.entities.Thread.create({
+        id: 100,
+        model: 'res.partner',
+    });
+    const follower = this.env.entities.Follower.create({
+        channel_id: 1,
+        email: "bla@bla.bla",
+        id: 2,
+        is_active: true,
+        is_editable: true,
+        name: "François Perusse",
+        partner_id: null,
+    });
+    const followerSubtype = this.env.entities.FollowerSubtype.create({
+        default: true,
+        internal: false,
+        id: 1,
+        name: "Dummy test",
+        res_model: 'res.partner'
+    });
+    follower.link({
+        subtypes: followerSubtype,
+        thread,
+    });
+    await this.createFollowerSubtypeComponent({
+        follower,
+        followerSubtype,
+    });
+    assert.strictEqual(
+        document.querySelectorAll('.o_FollowerSubtype').length,
+        1,
+        "should have follower subtype component"
+    );
+    assert.strictEqual(
+        document.querySelectorAll('.o_FollowerSubtype_label').length,
+        1,
+        "should have a label"
+    );
+    assert.strictEqual(
+        document.querySelectorAll('.o_FollowerSubtype_checkbox').length,
+        1,
+        "should have a checkbox"
+    );
+    assert.strictEqual(
+        document.querySelector('.o_FollowerSubtype_label').textContent,
+        "Dummy test",
+        "should have the name of the subtype as label"
+    );
+    assert.notOk(
+        document.querySelector('.o_FollowerSubtype_checkbox').checked,
+        "checkbox should not be checked as follower subtype is not followed"
+    );
+});
+
+QUnit.test('toggle follower subtype checkbox', async function (assert) {
+    assert.expect(5);
+
+    await this.start();
+    const thread = this.env.entities.Thread.create({
+        id: 100,
+        model: 'res.partner',
+    });
+    const follower = this.env.entities.Follower.create({
+        channel_id: 1,
+        email: "bla@bla.bla",
+        id: 2,
+        is_active: true,
+        is_editable: true,
+        name: "François Perusse",
+        partner_id: null,
+    });
+    const followerSubtype = this.env.entities.FollowerSubtype.create({
+        default: true,
+        internal: false,
+        id: 1,
+        name: "Dummy test",
+        res_model: 'res.partner'
+    });
+    follower.link({
+        thread,
+        subtypes: followerSubtype,
+    });
```

Comment:
I really don't like this copy/paste. This hides what is actually different in each test and what makes each of them unique.
Not no mention the pain to maintain and all other issues related to copy/pasting code.

Reply:
I agree but I know that @alexkuhn prefers having all needed data defined in the test to avoid looking after them. What would you suggest ?

Reply:
Well... Let's say we still need to find an agreement on this.

For consistency with the rest of code and to move forward with the branch, let's keep it as it is for now.

Reply:
Much prefer copy-pasting than sharing for data among tests.
I think this is a mistake to attempt defining tests by defining minimal change on data relatively to others. In my opinion, a test should be self-sustained and be fully understandable without requiring to jump somewhere else (e.g. find appropriate data in `beforeEach()` or utils function, move to mockRPC, understand magic value like `3` being the hard-coded global partner ID in tests, etc.).

So each test should looks like this:
```
"this should do that"
{
   [<make server data>]
   <make env>
   [<make entities>]
   <make component(s)>
   <user actions #1>
   <assertions #1>
   <user actions #2>
   <assertions #2>
   ...
}
```
If I start with knowing nothing on assumed data/components and can read and understand this test from top to bottom without needing "jumps", then this is a great test.

Reply:
This debate is for another PR, but I think we can achieve something similar to what we both want:

Just need to get rid of all the magic coming from `beforeEach`, `start` and more generally any use of `this` in tests. I think the main issue we have is everything happening silently there or with very unclear side effects.

It's good if tests become more explicit as you suggest here, but that setup inside each test should use appropriate helpers to avoid duplicate code.

Those helpers should prevent copy/paste, but also give the appropriate abstraction for the experienced reader: the reader without any knowledge can follow them and find the info there (no magic), the one with knowledge can ignore them and just see what is important for the test.